toolkit/tools/imagegen/diskutils/diskutils.go

```diff
@@ -660,3 +722,78 @@ func getPartUUID(device string) (uuid string, err error) {
 	uuid = strings.TrimSpace(stdout)
 	return
 }
+
+func checkSectorSizeFile(sectorFile string) (sectorSize uint64, err error) {
+	if exists, ferr := file.PathExists(sectorFile); ferr != nil {
+		logger.Log.Errorf("Error accessing sector size file %s", sectorFile)
+		err = ferr
+		return
+	} else if !exists {
+		err = fmt.Errorf("could not find the hw sector size file %s to obtain the sector size of the system", sectorFile)
+		return
+	}
+
+	fileContent, err := file.ReadLines(sectorFile)
+	if err != nil {
+		logger.Log.Errorf("Failed to read from %s: %s", sectorFile, err)
+		return
+	}
+
+	// sector file should only have one line, return error if not
+	if len(fileContent) != 1 {
+		err = fmt.Errorf("%s has more than one line", sectorFile)
+		return
+	}
+
+	sectorSize, err = strconv.ParseUint(fileContent[0], 10, 64)
+	return
+}
+
+func getSectorSize(diskDevPath string) (logicalSectorSize, physicalSectorSize uint64, err error) {
+	// Grab the specific disk name from /dev/xxx
+	diskName := diskDevPath[5:len(diskDevPath)]
+
+	hw_sector_size_file := fmt.Sprintf("/sys/block/%s/queue/hw_sector_size", diskName)
+	physical_sector_size_file := fmt.Sprintf("/sys/block/%s/queue/physical_block_size", diskName)
+
+	logicalSectorSize, err = checkSectorSizeFile(hw_sector_size_file)
+	if err != nil {
+		return
+	}
+
+	physicalSectorSize, err = checkSectorSizeFile(physical_sector_size_file)
+	return
+}
+
+func alignSectorAddress(sectorAddr, physicalSectorSize uint64) (alignedSector uint64) {
+	if sectorAddr%physicalSectorSize == 0 || sectorAddr < physicalSectorSize {
```

Comment:
Since the goal of this function is to align to physicalSectorSize, why is it okay for the `sectorAddr < physicalSectorSize` case to return a not aligned value?

Reply:
this condition is only used for the very first partition, which starts at 1MiB. Agreed that this logic is misleading and needs to be fixed.

Reply:
Refined the logic for this specific case.